File: base-programs/heat/go/src/heat.go
package heat

import (
	"gonum.org/v1/gonum/mat"
	"time"
	"math"
	"sync"
	"runtime"
	"strconv"
)

type HeatSrc struct {
	posX, posY, size, temp float64
}

func swap(x *mat.Dense, y *mat.Dense) {
	var temp mat.Dense
	temp = *x    /* save the value at address x */
	*x = *y      /* put y into x */
	*y = temp    /* put temp into y */
 }

func RunHeat(size string, threads_in int) string {

	dataSize := size
	iterations := 10
	N := 0
	if dataSize == "S" {
		N = 2200
	}
	if dataSize == "M" {
		N = 5200
	}
	if dataSize == "L" {
		N = 15200
	}

	heatSrcArr := []HeatSrc {
		HeatSrc{
			posX: 0.0,
			posY: 0.0,
			size: 1.0,
			temp: 1.0,
		},
		HeatSrc{
			posX: 1.0,
			posY: 1.0,
			size: 1.0,
			temp: 0.5,
		},
	}
	np := N + 2
	var wg sync.WaitGroup
	threads := 0
	threads = threads_in
	if runtime.NumCPU() < threads {
		threads = runtime.NumCPU()
	}
	start := time.Now()
	// Matrix init
	A := mat.NewDense(np, np, nil)
	A_help := mat.NewDense(np,np, nil)
	for i := 0; i < threads; i++ {
		wg.Add(1)
		go func (index int) {
			lowerBound := index * (np / threads)
			upperBound := (index + 1) * (np / threads)
			// fmt.Println(lowerBound, upperBound)
			// Init heatSrc
			for i, src := range heatSrcArr {
				// Top row
				dummy := i
				dummy++
				for j := lowerBound; j < upperBound; j++ {
					dist := math.Sqrt(math.Pow(float64(j) / float64(np - 1) - src.posX, 2) + math.Pow(src.posY, 2))
					if dist <= src.size {
						val := A.At(0, j) + (src.size - dist) / src.size * src.temp
						A.Set(0, j, val)
					}
				}
				// Bottom row
				for j := lowerBound; j < upperBound; j++ {
					dist := math.Sqrt(math.Pow(float64(j) / float64(np - 1) - src.posX, 2) + math.Pow(1 - src.posY, 2))
					if dist <= src.size {
						val := A.At(np - 1, j) + (src.size - dist) / src.size * src.temp
						A.Set(np - 1, j, val)
					}
				}
				// Leftmost column
				for j := lowerBound + 1; j < upperBound - 1; j++ {
					dist := math.Sqrt(math.Pow(src.posX, 2) + math.Pow(float64(j) / float64(np - 1) - src.posY, 2))
					if dist <= src.size {
						val := A.At(j, 0) + (src.size - dist) / src.size * src.temp
						A.Set(j, 0, val)
					}
				}
				// Rightmost column
				for j := lowerBound + 1; j < upperBound - 1; j++ {
					dist := math.Sqrt(1 - math.Pow(src.posX, 2) + math.Pow(float64(j) / float64(np - 1) - src.posY, 2))
					if dist <= src.size {
						val := A.At(j, np - 1) + (src.size - dist) / src.size * src.temp
						A.Set(j, np - 1, val)
					}
				}
			}
			wg.Done()
		}(i)
	}
	wg.Wait()
	// fmt.Print(A)
	//A_help := mat.DenseCopyOf(A)
	// A_help.Set(0, 0, 5.0)
	residualSum := 0.0
	residuals := make([]float64, threads)
	for iter := 0; iter < iterations; iter++ {
		for i := 0; i < threads; i++ {
			wg.Add(1)
			go func (index int) {
				lowerBound := index * (N / threads) + 1
				upperBound := (index + 1) * (N / threads) + 1
				val, diff, residual_thread := 0.0, 0.0, 0.0
				// fmt.Println(lowerBound, upperBound)
				for row := lowerBound; row < upperBound; row++ {
					for col := 1; col < np - 1; col++ {
						val = 0.25 * (A.At(row - 1, col) + A.At(row + 1, col) + A.At(row, col - 1) + A.At(row, col + 1))
						diff = val - A.At(row, col)
						A_help.Set(row, col, val)
						residual_thread += diff * diff
					}
				}
				residuals[index] = residual_thread
				wg.Done()
			}(i)
		}
		wg.Wait()
		residualSum = 0.0
		for i := range residuals {
			residualSum += residuals[i]
			residuals[i] = 0
		}
		swap(A, A_help)
	}

	duration := time.Since(start)

	timeInMS := float64(duration) / 1000000
	runtime.GC()
	return "Execution Time: " + strconv.FormatFloat(timeInMS, 'f', 6, 64) + "ms \nThreads: " + strconv.Itoa(threads)
}

File: gcr/heat/Java/src/main/java/heat/Heat.java
package heatpackage;

import java.util.Arrays;
import java.lang.Math;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.Vector;
import java.lang.Thread;
import java.util.Random;
import java.util.concurrent.TimeUnit;

public class Heat {

   public class HeatSrc {
      public double posX;
      public double posY;
      public double range;
      public double temp;

      public HeatSrc(double x, double y, double r, double t) {
         this.posX = x;
         this.posY = y;
         this.range = r;
         this.temp = t;
      }
   }

   public String runHeat(String input_in, Integer threads_in) throws InterruptedException {
      String input = input_in;

      System.out.println("Input: " + input);
      int N = 0;
      int iterations = 10;
      switch (input) {
         case "S":
            N = 200;
            break;
         case "M":
            N = 5200;
            break;
         case "L":
            N = 15200;
            break;
         default:
            break;
      }
      int threads = threads_in;
      if (threads > Runtime.getRuntime().availableProcessors()) {
         threads = Runtime.getRuntime().availableProcessors();
      }

      // double[] A_help = new double [N * N];

      // Initiate Heatsrces
      HeatSrc[] heatSrces = new HeatSrc[2];
      heatSrces[0] = new HeatSrc(0.0, 0.0, 1.0, 1.0);
      heatSrces[1] = new HeatSrc(1.0, 1.0, 1.0, 0.5);

      int np = N + 2;
      double[] A = new double[np * np];
      double[] A_help = new double[np * np];
      // double dist;

      ExecutorService executorService = Executors.newFixedThreadPool(threads);
      Vector<Integer> countInit = new Vector<Integer>(0);
      for (int t = 0; t < threads; t++) {
         // declare variables as final which will be used in method run below
         final int lowerBound = t * (np / threads);
         final int upperBound = (t + 1) * (np / threads);
         final int offset = np;
         // final float[] traces = new float[4];
         final int index = t;
         final double[] A_ref = A;
         final double[] A_ref_help = A_help;
         executorService.submit(new Runnable() {
            @Override
            public void run() {
               double dist;
               for (int i = 0; i < heatSrces.length; i++) {
                  // Top Row
                  for (int j = lowerBound; j < upperBound; j++) {
                     dist = Math.sqrt(Math.pow((double) j / (double) (np - 1) - heatSrces[i].posX, 2)
                           + Math.pow(heatSrces[i].posY, 2));
                     if (dist <= heatSrces[i].range) {
                        A_ref[j] += (heatSrces[i].range - dist) / heatSrces[i].range * heatSrces[i].temp;
                     }
                  }

                  /* bottom row */
                  for (int j = lowerBound; j < upperBound; j++) {
                     dist = Math.sqrt(Math.pow((double) j / (double) (np - 1) - heatSrces[i].posX, 2)
                           + Math.pow(1 - heatSrces[i].posY, 2));

                     if (dist <= heatSrces[i].range) {
                        A_ref[(np - 1) * np + j] += (heatSrces[i].range - dist) / heatSrces[i].range * heatSrces[i].temp;
                     }
                  }

                  /* leftmost column */
                  for (int j = lowerBound + 1; j < upperBound - 1; j++) {
                     dist = Math.sqrt(Math.pow(heatSrces[i].posX, 2)
                           + Math.pow((double) j / (double) (np - 1) - heatSrces[i].posY, 2));

                     if (dist <= heatSrces[i].range) {
                        A_ref[j * np] += (heatSrces[i].range - dist) / heatSrces[i].range * heatSrces[i].temp;
                     }
                  }

                  /* rightmost column */
                  for (int j = lowerBound + 1; j < upperBound - 1; j++) {
                     dist = Math.sqrt(Math.pow(1 - heatSrces[i].posX, 2)
                           + Math.pow((double) j / (double) (np - 1) - heatSrces[i].posY, 2));

                     if (dist <= heatSrces[i].range) {
                        A_ref[j * np + (np - 1)] += (heatSrces[i].range - dist) / heatSrces[i].range * heatSrces[i].temp;
                     }
                  }
               }

               for (int copy = 0; copy < A_ref.length; copy++) {
                  A_ref_help[copy] = A_ref[copy];
               }
               countInit.add(0);
               return;
            }
         });
      }
      while (countInit.size() < threads) {
         try {
            Thread.sleep(0);
         } catch (InterruptedException e) {
            System.out.println("Thread is interrupted");
         }
      }
      // double[] A_help = A.clone();
      long execTime = System.nanoTime();
      /*
       * for(double a:A) { System.out.print(" " + a + " "); }
       */

      // Arrays.fill(data, 0, data.length, generator.nextFloat());

      // float trace = 0;

      final double[] residuals = new double[threads];
      double residualSum = 0.0;
      // number of threads
      for (int iter = 0; iter < iterations; iter++) {
         Vector<Integer> count = new Vector<Integer>(0);
         for (int i = 0; i < threads; i++) {
            // declare variables as final which will be used in method run below
            final int lowerBound = i * (N / threads) + 1;
            final int upperBound = (i + 1) * (N / threads) + 1;
            final int offset = np;
            // final float[] traces = new float[4];
            final int index = i;
            final double[] A_ref = A;
            final double[] A_ref_help = A_help;
            executorService.submit(new Runnable() {
               @Override
               public void run() {
                  double diff, val;
                  double resiudal_thread = 0;
                  for (int row = lowerBound; row < upperBound; row++) {
                     for (int col = 1; col < offset - 1; col++) {
                        // x_help[row] += A[row * offset + col] * x[col];
                        val = 0.25 * (A_ref[row * offset + col - 1] + A_ref[row * offset + col + 1]
                              + A_ref[(row - 1) * offset + col] + A_ref[(row + 1) * offset + col]);
                        diff = val - A_ref[row * offset + col];
                        A_ref_help[row * offset + col] = val;
                        resiudal_thread += diff * diff;
                     }
                  }
                  residuals[index] = resiudal_thread;
                  count.add(0);
                  return;
               }
            });
         }
         while (count.size() < threads) {
            try {
               Thread.sleep(0);
            } catch (InterruptedException e) {
               System.out.println("Thread is interrupted");
            }
         }
         // Add residuals
         residualSum = 0.0;
         for (int res = 0; res < residuals.length; res++) {
            residualSum += residuals[res];
            residuals[res] = 0.0;
         }
         double[] temp = A;
         A = A_help;
         A_help = temp;
      }
      // System.out.println(residualSum);
      executorService.shutdown();
      execTime = System.nanoTime() - execTime;
      try {
         if (!executorService.awaitTermination(5000, TimeUnit.MILLISECONDS)) {
            executorService.shutdownNow();
         }
      } catch (InterruptedException e) {
         executorService.shutdownNow();
      }
      return "Execution time: " + (double) execTime / 1000000 + "ms, Threads: " + threads;
   }
}